### PR TITLE
Copy Post: Add a metadata whitelist

### LINKED ISSUE #14840: Editor: After Copy a Post, Publicize fails
Copy-and-pasting the original report by @KokkieH:

### Summary:

When using the copy post feature in Calypso, the new post does not trigger Publicize if the post being copied has already publicized.

### To replicate:

1. Go to https://wordpress.com/sharing and ensure the site has an active Publicize connection to any service.
2. Go to https://wordpress.com/post and create a new post. Add some content and publish it.
3. Create another new post. In Post Settings, click More Options and select Copy Post. Select the previous post.
4. Make changes to the title and content of the new post, as a user would do when using an old post as a template for a new one.
5. Under Sharing in Post Settings, make sure Publicize is enabled, and publish the post.

### What I expected:

That the second post would Publicize to the connected social media account the same as the first, as it's a new post that has never been Publicized before.

### What happened instead:

The second post was not Publicized. Looking at the site's RC it indicates that Publicize failed because it detected the post had already been Publicized, even though this was a brand new post.

As a control I also used the Copy Post feature via WP-Admin, and in that case the post did Publicize correctly.

### PR DESCRIPTION
**Pull request by Copons**
Fixes #14840

Adds a metadata whitelist in order to accurately control which metadata are to be copied over with the Copy Post feature.

The whitelist currently only contains `geo_latitude` and `geo_longitude`, which can be safely copied.
Most notably excluded from the whitelist are the Publicize-related metadata, that caused the aforementioned issue.

### CODE REVIEW

**Comment by nylen** [comment]
This seems like it could be written more cleanly:

```js
actions.updateMetadata(
	// Whitelist copied meta keys to avoid issues such as Publicize not triggering
	// on the copied post (https://github.com/Automattic/wp-calypso/issues/14840).
	pick( postToCopy.metadata, [
		'geo_latitude',
		'geo_longitude',
	] )
);
```

**Comment by Copons** [comment]
@nylen unfortunately that wouldn't work because `_.pick()` expects an object, but `post.metadata` is an array of objects.

It's a bit counterintuitive, but it goes pretty much like this.

When we _retrieve_ `post.metadata` it's an array of objects:
```es6
[
  { key: 'foo', value: 'bar' },
  { key: 'baz', value: 'qux' },
]
```

When we _update_ it, though, we need to send a flat object:
```es6
{
  foo: 'bar',
  baz: 'qux',
}
```

This conversion is performed via the `reduce()` accumulating into `{}`.

Actually, I could `reduce` the array into an object **before**, and `pick` from the whitelist **after**.
I guess from a performance standpoint the differences are minimal, but I agree that it's cleaner and more readable.

**Comment by nylen** [comment]
Ah ok, I missed that, thanks for the clarification. Another option would be to do the whitelist inside of the `reduce` call (via a `switch` statement or `includes`) but now the code is looking pretty clean either way.

File: client/post-editor/controller.js [hunk 1]
```diff
@@ -158,12 +158,27 @@ function startEditingPostCopy( siteId, postToCopyId, context ) {
...
+		 *
+		 * This is needed because blindly passing all post metadata to `updateMetadata()`
+		 * causes unforeseeable issues, such as Publicize not triggering on the copied post.
+		 *
+		 * @see https://github.com/Automattic/wp-calypso/issues/14840
+		 */
+		const metadataWhitelist = [
+			'geo_latitude',
+			'geo_longitude',
+		];
```
Comment (gwwar) [inline client/post-editor/controller.js]: Are geo coordinates interesting to copy over?

Comment (Copons) [inline client/post-editor/controller.js]: I don't honestly have a strong opinion about it, but we discussed about it in the issue (https://github.com/Automattic/wp-calypso/issues/14840#issuecomment-306907743).

**Pull request merged**